### PR TITLE
Improve programmer productivity with diagnostics

### LINKED ISSUE #299: Source locations in errors follow-up
Follow-up following #295:

- [ ] Take the target directly in throwDiagnostic and don't require programmer to call getSourceLocationOfXxx
- [ ] Harden better against a node without a parent. Preserve the original diagnostic along with ICE.

### PR DESCRIPTION
**Pull request by nguerrera**
Fix #299, doing follow-up to #295

1. Make throwDiagnostic take the target directly and figure out the location rather than having the programmer call getSourceLocation.

2. If getting location or formatting message fails, include as much of the diagnostic as possible in the internal compiler error.

Here is how the bug fixed by #300 manifests now:

```
ADL compiler v0.2.1

Internal compiler error!
File issue at https://github.com/azure/adl

Error: <unknown location>:1:1 - error: Unknown identifier T
    at formatDiagnostic (file:///C:/Src/adl/packages/adl/dist/compiler/diagnostics.js:61:15)
    at throwDiagnostic (file:///C:/Src/adl/packages/adl/dist/compiler/diagnostics.js:32:31)
    at resolveIdentifier (file:///C:/Src/adl/packages/adl/dist/compiler/checker.js:282:13)
    at checkIdentifier (file:///C:/Src/adl/packages/adl/dist/compiler/checker.js:259:25)
    at getTypeForNode (file:///C:/Src/adl/packages/adl/dist/compiler/checker.js:79:24)
    at file:///C:/Src/adl/packages/adl/dist/compiler/checker.js:355:34
    at Array.map (<anonymous>)
    at checkClassHeritage (file:///C:/Src/adl/packages/adl/dist/compiler/checker.js:354:25)
    at checkModel (file:///C:/Src/adl/packages/adl/dist/compiler/checker.js:310:19)
    at getTypeForNode (file:///C:/Src/adl/packages/adl/dist/compiler/checker.js:70:24)

Error: Cannot obtain source file of unbound node.
    at getSourceLocationOfNode (file:///C:/Src/adl/packages/adl/dist/compiler/diagnostics.js:85:15)
    at getSourceLocation (file:///C:/Src/adl/packages/adl/dist/compiler/diagnostics.js:77:40)
    at formatDiagnostic (file:///C:/Src/adl/packages/adl/dist/compiler/diagnostics.js:42:20)
    at throwDiagnostic (file:///C:/Src/adl/packages/adl/dist/compiler/diagnostics.js:32:31)
    at resolveIdentifier (file:///C:/Src/adl/packages/adl/dist/compiler/checker.js:282:13)
    at checkIdentifier (file:///C:/Src/adl/packages/adl/dist/compiler/checker.js:259:25)
    at getTypeForNode (file:///C:/Src/adl/packages/adl/dist/compiler/checker.js:79:24)
    at file:///C:/Src/adl/packages/adl/dist/compiler/checker.js:355:34
    at Array.map (<anonymous>)
    at checkClassHeritage (file:///C:/Src/adl/packages/adl/dist/compiler/checker.js:354:25)
```

And, if we add a second easy to cause bug on the same line of code by adding a format placeholder to the message and neglecting to pass an arg for it:

```
ADL compiler v0.2.1

Internal compiler error!
File issue at https://github.com/azure/adl

Error: <unknown location>:1:1 - error: Unknown <missing argument> identifier T
    at formatDiagnostic (file:///C:/Src/adl/packages/adl/dist/compiler/diagnostics.js:61:15)
    at throwDiagnostic (file:///C:/Src/adl/packages/adl/dist/compiler/diagnostics.js:32:31)
    at resolveIdentifier (file:///C:/Src/adl/packages/adl/dist/compiler/checker.js:282:13)
    at checkIdentifier (file:///C:/Src/adl/packages/adl/dist/compiler/checker.js:259:25)
    at getTypeForNode (file:///C:/Src/adl/packages/adl/dist/compiler/checker.js:79:24)
    at file:///C:/Src/adl/packages/adl/dist/compiler/checker.js:355:34
    at Array.map (<anonymous>)
    at checkClassHeritage (file:///C:/Src/adl/packages/adl/dist/compiler/checker.js:354:25)
    at checkModel (file:///C:/Src/adl/packages/adl/dist/compiler/checker.js:310:19)
    at getTypeForNode (file:///C:/Src/adl/packages/adl/dist/compiler/checker.js:70:24)

Error: Cannot obtain source file of unbound node.
    at getSourceLocationOfNode (file:///C:/Src/adl/packages/adl/dist/compiler/diagnostics.js:85:15)
    at getSourceLocation (file:///C:/Src/adl/packages/adl/dist/compiler/diagnostics.js:77:40)
    at formatDiagnostic (file:///C:/Src/adl/packages/adl/dist/compiler/diagnostics.js:42:20)
    at throwDiagnostic (file:///C:/Src/adl/packages/adl/dist/compiler/diagnostics.js:32:31)
    at resolveIdentifier (file:///C:/Src/adl/packages/adl/dist/compiler/checker.js:282:13)
    at checkIdentifier (file:///C:/Src/adl/packages/adl/dist/compiler/checker.js:259:25)
    at getTypeForNode (file:///C:/Src/adl/packages/adl/dist/compiler/checker.js:79:24)
    at file:///C:/Src/adl/packages/adl/dist/compiler/checker.js:355:34
    at Array.map (<anonymous>)
    at checkClassHeritage (file:///C:/Src/adl/packages/adl/dist/compiler/checker.js:354:25)

Error: Missing format argument.
    at file:///C:/Src/adl/packages/adl/dist/compiler/diagnostics.js:98:21
    at String.replace (<anonymous>)
    at format (file:///C:/Src/adl/packages/adl/dist/compiler/diagnostics.js:95:26)
    at formatDiagnostic (file:///C:/Src/adl/packages/adl/dist/compiler/diagnostics.js:56:32)
    at throwDiagnostic (file:///C:/Src/adl/packages/adl/dist/compiler/diagnostics.js:32:31)
    at resolveIdentifier (file:///C:/Src/adl/packages/adl/dist/compiler/checker.js:282:13)
    at checkIdentifier (file:///C:/Src/adl/packages/adl/dist/compiler/checker.js:259:25)
    at getTypeForNode (file:///C:/Src/adl/packages/adl/dist/compiler/checker.js:79:24)
    at file:///C:/Src/adl/packages/adl/dist/compiler/checker.js:355:34
    at Array.map (<anonymous>)
```

### CODE REVIEW

File: packages/adl/compiler/diagnostics.ts [hunk 1]
```diff
@@ -11,48 +14,119 @@ export class DiagnosticError extends Error {
   }
 }
 
-export type ErrorHandler = (message: Message | string, location: SourceLocation, ...args: Array<string | number>) => void;
+export class ChainedError extends Error {
+  readonly innerErrors: readonly Error[];
```
Comment (nguerrera) [inline packages/adl/compiler/diagnostics.ts:18]: Does something like this exist in JS already? Couldn't find it.

File: packages/adl/compiler/diagnostics.ts [hunk 2]
```diff
@@ -11,48 +14,119 @@ export class DiagnosticError extends Error {
...
+
+  constructor(message: string, ...innerErrors: (Error | undefined)[]) {
+    super(message);
+    this.innerErrors = innerErrors.filter(e => e !== undefined) as Error[];
```
Comment (nguerrera) [inline packages/adl/compiler/diagnostics.ts:22]: This cast annoys me. Is there a way to remove it without implementing a single-use type predicate?

**Review by daviwil — APPROVED** [review]
Looks great!  This will definitely make it easier to use these functions.

File: packages/adl/compiler/diagnostics.ts [hunk 3]
```diff
@@ -11,48 +12,126 @@ export class DiagnosticError extends Error {
...
+ */
+export class ChainedError extends Error {
+  readonly innerErrors: readonly Error[];
+
+  constructor(message: string, ...innerErrors: (Error | undefined)[]) {
+    super(message);
+    this.innerErrors = innerErrors.filter(isNotUndefined);
+  }
+
+  dumpInnerStacks(writeLine: (s: string) => void) {
```
Comment (nguerrera) [inline packages/adl/compiler/diagnostics.ts:26]: We don't have the recursive case yet, but this misses the outer error of inner chained error. I will fix this.

Comment (nguerrera) [reply packages/adl/compiler/diagnostics.ts:26]: This was wrong, it is correct, but I'm writing it a nicer way, still. LOL.

**Pull request merged**